Add migration guide for react hooks for ably-js v2 (#1698)
ably/ably-js: docs/react-migration-guide.md

# React hooks upgrade / migration guide

> ## Version ?? to ??

Author: Previous iterations of this migration guide used versions used by separate react-hooks library, which was not part of ably-js repo (see line https://github.com/ably/ably-js/pull/1698/files#diff-f8ce6205300ecf411be3e91af62c0ecbaf88971e7cd082fc9193ef20cdea32eeR205), so previous guide was for `Version 2.x to 3.x`.

What do we want write now here?
React Hooks are now part of ably-js and follow its versioning.

Reviewer: the "2.x to 3.x" thing shouldn't have been there in the first place (probably my bad) - we were originally making some breaking changes as part of an `@ably-labs/react-hooks` v3 major release but instead decided to merge the hooks into the core `ably` package so there never was a 3.x release in the end.

I think we should change the header of the previous migration guide to "`@ably-labs/react-hooks` to 1.x" and then the new migration guide can simply be labelled "1.x to 2.x"

Author: how about "`@ably-labs/react-hooks` to `ably-js 1.x`"? so it's a bit more descriptive that we moved to `ably-js` repo

Reviewer: sounds good to me, but i would go for `ably` rather than `ably-js` since that's the name of the npm package

Author: yea, makes sence! will do `ably`

Author: changed in https://github.com/ably/ably-js/commit/1c5d80da35f5ca560afb988b4df7cfd7ce23d6ef

# React hooks upgrade / migration guide

## Version 1.x to 2.x

### ChannelProvider

> In previous versions, you were able to provide channel options as a parameter in the `useChannel`/`usePresence` hooks. This often led to errors when using hooks with the same channel name but different options, or when attempting to dynamically change options for a channel.

Reviewer: i think slightly softer language might make this sound less scary 😅 
```suggestion
In previous versions, you were able to provide channel options as a parameter in the `useChannel`/`usePresence` hooks. This could, in some cases, lead to errors when using hooks with the same channel name but different options, or when attempting to dynamically change options for a channel.
```

Author: Agree 😅

  { foo: 'bar' },
);
```

With this:

```jsx
// in a parent component:
return <ChannelProvider channelName="presence-channel-name" options={{ modes: ['PRESENCE'] }}>
  {children}
</ChannelProvider>

// in a child component:
const { updateStatus } = usePresence({ channelName: 'presence-channel-name' }, { foo: 'bar' });
```

Additionally, if you were calling `.setOptions()` on a channel instance returned by the `useChannel` hook before, you must remove those calls and instead modify options provided to the `ChannelProvider` component if you want to change channel options during runtime.

Change this:

```jsx
const { channel } = useChannel({ channelName: 'your-channel-name' }, (message) => {
  console.log(message);
});
channel.setOptions({ params: { rewind: '1' } });
```

To this:

```jsx
// in a parent component:
// change channel options during runtime using state
const [options, setOptions] = useState({});
return <ChannelProvider channelName="your-channel-name" options={options}>
  {children}
</ChannelProvider>

// in a child component:
const { channel } = useChannel({ channelName: 'your-channel-name' }, (message) => {
  console.log(message);

> });

Reviewer: ```suggestion
});

// use the useState setter to change the options provided to the `ChannelProvider` component
setOptions({ params: { rewind: '1' } });
```

```

### `usePresence` is split into two hooks

The functionality of the `usePresence` hook has been split into two separate hooks.

The `usePresence` hook can now only be used to enter the presence with optional initial state and update the presence status for the current client. It no longer returns the `presenceData` value and does not accept the `onPresenceUpdated` callback as its third parameter.

To listen for presence updates, a new hook called `usePresenceListener` has been introduced. This hook returns the `presenceData` object previously returned by `usePresence` and accepts an `onPresenceMessageReceived` callback as its second parameter, which is called on new presence messages.

Replace this:

```jsx
const { presenceData, updateStatus } = usePresence(
  { channelName: 'presence-channel-name' },
  { foo: 'bar' },
  (update) => {
    console.log(update);
  },
);
```

With this:

```jsx
const { updateStatus } = usePresence({ channelName: 'presence-channel-name' }, { foo: 'bar' });
const { presenceData } = usePresenceListener({ channelName: 'presence-channel-name' }, (update) => {
  console.log(update);
});
```

> ### Renaming `id` to `ablyId`

Author: ```suggestion
### Renaming optional `id` field to `ablyId`
```

> All instances of the `id` field, which were used to specify the `AblyProvider` component and the underlying Ably client to use, have been renamed to `ablyId`.

Reviewer: I think it would be nice to make it clear that this was an optional feature, i'd bet the vast majority of users aren't using more than one provider so this section might be a bit confusing to them
```suggestion
All instances of the `id` field, which optionally were used to specify the `AblyProvider` component and the underlying Ably client to use, have been renamed to `ablyId`.
```

Author: True, will also change the header to `Renaming optional "id" field to "ablyId"`
